File: 2021/3b.jl
function run()
    t = time()
    lines = open("input_3.txt") do f
        [parse(Int, l, base=2) for l in readlines(f)]
    end

    val = ratingCalc(lines, "oxy")
    val *= ratingCalc(lines, "co2")
    println("🎄🎅 Part 2: ", val)
    println(time()-t)
end

function mostCommon(col::Int, inputs::Vector{Int}, type::String)
    n = count([i & 2^col == 2^col for i in inputs])
    
    if type == "oxy"
        if n >= length(inputs)//2
            return 2^col
        else
            return 0
        end
    else
        if n < length(inputs)//2
            return 2^col
        else
            return 0
        end
    end
end

function ratingCalc(lines::Vector{Int}, t::String)
    rating = 0
    tmp_lines = lines
    col = 11
    while length(tmp_lines) > 1
        mc = mostCommon(col, tmp_lines, t)
        tmp_lines = [t for t in tmp_lines if t & 2^col == mc]
        col -= 1
    end
    
    return tmp_lines[1]
end

run()

File: 2022/3b.py
from utils import readInput

def ratingCalc(lines, t="oxy"):
    rating = ""
    tmp_lines = lines
    col = 0
    while len(tmp_lines) > 1:
        thrs = len(tmp_lines)/2
        counts = sum([int(l[col]) for l in tmp_lines])
        if t == "oxy":
            if counts >= thrs:
                res = "1"
            else:
                res = "0"
        elif t == "co2":
            if counts >= thrs:
                res = "0"
            else:
                res = "1"
        rating += res
    
        tmp_tmp_lines = [l for l in tmp_lines if l[col]==res]
        tmp_lines = tmp_tmp_lines
        col += 1
    else:
        if len(rating) != len(lines[0]):
            rating += tmp_lines[0][len(rating):]
    return rating

import time
tstart = time.time()
lines = readInput("input_3.txt")
oxygen = ratingCalc(lines, "oxy")
scrubber = ratingCalc(lines, "co2")
res = int(scrubber, 2)*int(oxygen, 2)

print ("🎄🎅 Part 2: ", res)
print (time.time()-tstart)
